### PR TITLE
Allow the user to provide a pairwise distance metric

### PR DESCRIPTION
**Pull request by Linux-cpp-lisp**
Adds an optional keyword argument, `distances`, through which a pairwise distance matrix can be provided by the user, bypassing the built-in Euclidean metric. This allows a user to, for example, use the many efficient metrics in [`sklearn.metrics.pairwise_distances`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html). 

Technically, I think a user could provide _only_ `distances` and not `points`, as `points` is only used to compute `distances`. This opens the door to using DPC on stranger point sets and with general similarity kernels, which is a nice flexibility to have.

### CODE REVIEW

**Comment by cwehmeyer** [comment]
Thanks!
